spark/src/main/scala/org/apache/spark/sql/delta/commands/UpdateCommand.scala

```diff
@@ -117,76 +118,153 @@ case class UpdateCommand(
     val (metadataPredicates, dataPredicates) =
       DeltaTableUtils.splitMetadataAndDataPredicates(
         updateCondition, txn.metadata.partitionColumns, sparkSession)
-    val candidateFiles = txn.filterFiles(metadataPredicates ++ dataPredicates)
+
+    // Should we write the DVs to represent updated rows?
+    val shouldWriteDeletionVectors = shouldWritePersistentDeletionVectors(sparkSession, txn)
+    val candidateFiles = txn.filterFiles(
+      metadataPredicates ++ dataPredicates,
+      keepNumRecords = shouldWriteDeletionVectors)
+
     val nameToAddFile = generateCandidateFileMap(deltaLog.dataPath, candidateFiles)
 
     scanTimeMs = (System.nanoTime() - startTime) / 1000 / 1000
 
-    val filesToRewrite: Seq[AddFile] = if (candidateFiles.isEmpty) {
+    val allActions: Seq[FileAction] = if (candidateFiles.isEmpty) {
       // Case 1: Do nothing if no row qualifies the partition predicates
       // that are part of Update condition
       Nil
     } else if (dataPredicates.isEmpty) {
       // Case 2: Update all the rows from the files that are in the specified partitions
       // when the data filter is empty
-      candidateFiles
+      numTouchedFiles = candidateFiles.size
+
+      // Rewrite all candidateFiles
+      val addAndRemoveActions =
+        withStatusCode("DELTA", UpdateCommand.rewritingFilesMsg(candidateFiles.size)) {
+          if (candidateFiles.nonEmpty) {
+            rewriteFiles(
+              sparkSession,
+              txn,
+              tahoeFileIndex.path,
+              candidateFiles,
+              nameToAddFile,
+              updateCondition,
+              generateRemoveFileActions = true,
+              copyUnmodifiedRows = true)
+          } else {
+            Nil
+          }
+        }
+
+      rewriteTimeMs = (System.nanoTime() - startTime) / 1000 / 1000 - scanTimeMs
```

Comment:
Use `TimeUnit` constants? (I know this was already in the old code...but we could make it nicer since touching it anyway).

Reply:
Yup, changed them all!